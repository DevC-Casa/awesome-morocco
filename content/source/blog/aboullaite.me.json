{
  "name": "Aboullaite blog",
  "url": "https://aboullaite.me/",
  "description": "React, javascript , React Native ...",
  "rssUrl": "https://aboullaite.me/rss/"
}
