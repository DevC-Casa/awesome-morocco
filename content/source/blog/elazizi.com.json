{
  "name": "El Azizi blog",
  "url": "https://elazizi.com",
  "description": "React, javascript , React Native ...",
  "rssUrl": "https://elazizi.com/rss.xml"
}
