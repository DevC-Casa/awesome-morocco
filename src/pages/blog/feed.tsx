import { type NextPage } from "next";
import Head from "next/head";

import { NavBar, PageBanner, TabsList } from "~/components";

const Home: NextPage = () => {
  return (
    <>
      <Head>
        <title>Blogs | Awesome Morocco</title>
        <meta name="description" content="Generated by create-t3-app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <NavBar />
      <PageBanner
        title="Blogs"
        description="Explore latest articles written moroccan tech community"
      />
      <main className="flex flex-1 flex-col items-center justify-center px-20 text-center">
        <TabsList
          routes={[
            { name: "Blogs", link: "/blog" },
            { name: "feed", link: "/blog/feed" },
          ]}
          activeIndex={1}
        />
        <Feed />
      </main>
    </>
  );
};

export default Home;

const Feed: React.FC = () => {
  return (
    <div className="grid gap-4 sm:grid-cols-2 sm:gap-8 lg:grid-cols-2 xl:grid-cols-2 xl:gap-8">
      feed
    </div>
  );
};
