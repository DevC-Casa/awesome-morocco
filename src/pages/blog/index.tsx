import { type NextPage } from "next";
import Head from "next/head";

import { api } from "~/utils/api";
import { NavBar, PageBanner, TabsList } from "~/components";

const Home: NextPage = () => {
  return (
    <>
      <Head>
        <title>Blogs | Awesome Morocco</title>
        <meta name="description" content="Generated by create-t3-app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <NavBar />
      <PageBanner
        title="Blogs"
        description="Explore latest articles written moroccan tech community"
      />
      <main className="flex flex-1 flex-col items-center justify-center px-20 text-center">
        <TabsList
          routes={[
            { name: "Blogs", link: "/blog" },
            { name: "feed", link: "/blog/feed" },
          ]}
          activeIndex={0}
        />
        <Blogs />
      </main>
    </>
  );
};

export default Home;

const Blogs: React.FC = () => {
  const { data, isLoading } = api.blog.blogs.useQuery();

  if (isLoading) {
    return <p>Loading...</p>;
  }

  return (
    <section className="w-screen py-20">
      <div className="mx-auto grid max-w-screen-lg grid-cols-1 gap-5 p-5 sm:grid-cols-2 md:grid-cols-3 lg:gap-10">
        {data?.map((blog) => (
          <Article
            key={blog.id}
            title={blog.title}
            description={blog.description}
            image={blog.image}
          />
        ))}
      </div>
    </section>
  );
};

type ArticleProps = {
  title: string;
  description: string;
  image: string;
};
const Article = ({ title, description, image }: ArticleProps) => {
  return (
    <div className="group relative m-10 h-60 w-full  cursor-pointer overflow-hidden rounded-lg shadow-md">
      <div className="absolute left-0 top-0 h-full w-full  transition-all duration-300 ease-in-out group-hover:-top-96">
        <img
          className="h-4/6 w-full object-cover"
          src={image}
          alt={`${title} website image`}
        />
        <div className="flex h-2/6 items-center justify-center  ">
          <h1 className=" px-4 text-center font-serif text-xl font-semibold">
            {title}
          </h1>
        </div>
      </div>

      <div className="absolute left-0 -bottom-96 flex h-full w-full flex-col justify-center transition-all duration-300 ease-in-out group-hover:bottom-0">
        <h1 className="mb-2 px-8 text-center font-serif text-xl font-semibold ">
          {title}
        </h1>
        <p className="px-8 text-center">{description}</p>
      </div>
    </div>
  );
};
