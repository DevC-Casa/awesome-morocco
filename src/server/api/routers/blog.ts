import { z } from "zod";

import {
  createTRPCRouter,
  protectedProcedure,
  publicProcedure,
} from "~/server/api/trpc";
import { TRPCError } from "@trpc/server";
import { getURLOpenGraphMetadata } from "~/server/utils/get-url-open-graph-metadata";

export const blogRouter = createTRPCRouter({
  new_blog: protectedProcedure
    .input(
      z.object({
        url: z.string().url(),
        rss: z.union([z.string().url().optional(), z.string().optional()]),
      })
    )
    .mutation(async ({ input, ctx }) => {
      const data = await getURLOpenGraphMetadata(input.url);
      if (data === null) {
        throw new TRPCError({
          code: "BAD_REQUEST",
          message: "Error getting metadata from blog url",
        });
      }

      console.log(data);

      const dataCreated = await ctx.prisma.blog.create({
        data: {
          url: input.url,
          title: data.title,
          description: data.description || "",
          image: data.image || "", // add placeholder image here in case
          rss: input.rss || "/images/placeholder.png",
          lastRSSUpdatedAt: new Date("2000-01-01T00:00:00.000Z"), // default value :)
          author: {
            connect: {
              id: ctx.session.user.id,
            },
          },
        },
      });

      return dataCreated;
    }),
  blogs: publicProcedure.query(async ({ ctx }) => {
    const data = await ctx.prisma.blog.findMany();
    return data;
  }),
  my_blogs: protectedProcedure.query(async ({ ctx }) => {
    const data = await ctx.prisma.blog.findMany({
      where: {
        authorId: ctx.session.user.id,
      },
    });
    return data;
  }),
});
