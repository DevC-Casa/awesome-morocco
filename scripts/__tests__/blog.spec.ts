import {
  EMPTY_LIST_ERROR,
  NO_ITEMS_ERROR,
  RSS_URL_ERROR,
} from "../src/constants";
import { feedPromise } from "../src/blog/feed-url-promise";
import { generatePostsFromRssUrls } from "../src/blog/generate-posts-from-rss-urls";
import { getErrorReporting } from "../src/blog/get-errors-reporting";

const BLOG_1 = {
  title: "blog 1",
  link: "http://blog1.me/",
  items: [
    {
      title: "Blog 1 Article title ",
      link: "http://blog1.me/article-1/",
      pubDate: "2020-01-08T13:04:30.000Z",
    },
    {
      title: "Blog 1 Article title 2",
      link: "http://blog1.me/article-2/",
      pubDate: "2020-01-09T13:04:30.000Z",
    },
  ],
};

const BLOG_2 = {
  title: "blog 2",
  link: "http://blog2.me/",
  items: [
    {
      title: "Blog 2  Article title 1",
      link: "http://blog2.me/article-1/",
      pubDate: "2020-01-10T13:04:30.000Z",
    },
    {
      title: "Blog 2 Article title 2",
      link: "http://blog2.me/article-2/",
      pubDate: "2020-01-12T13:04:30.000Z",
    },
  ],
};

const NO_ITEMS_BLOG = {
  title: "No Items Blog",
  link: "http://no-items-blog.me/",
};

jest.mock("../src/blog/get-rss-urls-list", () => ({
  getRssUrlsList: () => {
    return ["http://blog1.me/", "http://blog2.me/"];
  },
}));

jest.mock("../src/blog/parser", () => ({
  parseUrl: (url: string) => {
    const blog = [BLOG_1, BLOG_2, NO_ITEMS_BLOG].find(
      (blog) => blog.link === url
    );
    if (blog !== undefined) return Promise.resolve(blog);
    else throw new Error("error!");
  },
}));

describe("Blog posts generation from rss feeds ", () => {
  test("return error when list is empty", async () => {
    const { posts, logs } = await generatePostsFromRssUrls([]);
    expect(posts.length).toBe(0);
    expect(logs).toBe(EMPTY_LIST_ERROR);
  });

  test("reporting should return the correct message", () => {
    const logs = getErrorReporting(6, 30, [
      "Error message 1",
      "Error message 2",
    ]);
    expect(logs).toMatchSnapshot();
  });

  test("Return the correct items as posts from url ", async () => {
    const rss = await feedPromise(BLOG_1.link);
    expect(rss[0].link).toBe(BLOG_1.items[0].link);
    expect(rss[1].link).toBe(BLOG_1.items[1].link);
  });

  test("should return errors for no items rss", async () => {
    await expect(feedPromise(NO_ITEMS_BLOG.link)).rejects.toBe(NO_ITEMS_ERROR);
  });

  test("should return error for invalid rss links", async () => {
    const invalidURL = "https://myinvalidlink.com/";
    await expect(feedPromise(invalidURL)).rejects.toBe(
      `${invalidURL}: ${RSS_URL_ERROR}`
    );
  });

  test("Return the correct items as posts from url ", async () => {
    const { posts } = await generatePostsFromRssUrls([
      BLOG_1.link,
      BLOG_2.link,
    ]);
    expect(posts.length).toBe(4);
    expect(posts[0].link).toBe(BLOG_2.items[1].link);
    expect(
      posts[0].pubDate.valueOf() - posts[1].pubDate.valueOf()
    ).toBeGreaterThan(0);
  });

  test("return  valid posts and correct errors  for real use cases", async () => {
    const invalidURL = "https://myinvalidlink.com/";
    const { posts, errors } = await generatePostsFromRssUrls([
      BLOG_1.link,
      BLOG_2.link,
      invalidURL,
    ]);
    expect(posts.length).toBe(4);
    expect(errors.length).toBe(1);
    expect(errors[0]).toBe(`${invalidURL}: ${RSS_URL_ERROR}`);
  });
});
