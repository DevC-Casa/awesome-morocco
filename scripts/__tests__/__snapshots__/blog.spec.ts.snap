// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog posts generation from rss feeds  reporting should return the correct message 1`] = `
"
✅ 30 blog posts has been collected from 4 RSS feeds
❌ Error message 1
❌ Error message 2
      "
`;
